clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -183,9 +183,23 @@ public void assignFromSubscribed(Collection<TopicPartition> assignments) {
         Set<TopicPartition> newAssignment = new HashSet<>(assignments);
         removeAllLagSensors(newAssignment);
 
-        for (TopicPartition tp : assignments)
-            if (!this.subscription.contains(tp.topic()))
-                throw new IllegalArgumentException("Assigned partition " + tp + " for non-subscribed topic.");
+        if (this.subscribedPattern != null) {
```

Comment:
Can we add a unit test covering the cases where the assignment contains an unsubscribed partition which matches and doesn't match the regex?